[TASK] Improve introduction of plugins (#4727)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/ContentElements/Index.rst

..  contents::
    :local:

In TYPO3, content elements and plugins are both used to present and manage
content on a website, but they serve different purposes and have distinct
characteristics:

..  card-grid::
    :columns: 1
    :columns-md: 2
    :card-height: 100

    ..  card:: :ref:`Content elements <content-elements>`

        ..  card-image:: /Images/ManualScreenshots/Backend/content_element_example.png
            :alt: Content element Text & Media

        A content element is a standard unit for managing and displaying content,

>         such as text, images, videos, tables, and more.

Reviewer: ```suggestion
        such as text, images, videos, tables and more.
```

Author: This is a "oxford coma" and therefore needed

possible without writing PHP code and can therefore also be done by
TYPO3 integrators.

..  _plugins:

Plugins in TYPO3
----------------

> A **plugin** in TYPO3 is a more complex module, typically providing dynamic

Reviewer: A "module" is a special name for own implementations in TYPO3 backend. I would try to prevent using that word in FE context. What about implementation or integration?

```suggestion
A **plugin** in TYPO3 is a more complex implementation, typically providing dynamic
```

or interactive functionality. Plugins are usually provided by extensions
that introduce new features to the website.

> The data to be displayed is usually supplied by a special PHP class
> called a "controller". Depending on the technology used in the controller
> the plugin can be an Extbase plugin or a plain plugin.

Reviewer: There are no controllers in abstract plugin implementation.

```suggestion
The data to be displayed can be realized by an Extbase plugin 
following the MVC pattern or with help of a very basic plugin 
structure called "Abstract Plugin" where you have full access to
TYPO3 API.
```

Author: The AbstractPlugin rests in peace: https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.4/Deprecation-100639-DeprecateAbstractPlugin.html#deprecation-100639-deprecate-abstractplugin and you should implement a Controller pattern in non-extbase plugins, though noone can force you to...

..  _plugins-extbase:

Extbase plugins
~~~~~~~~~~~~~~~

For usage in the TYPO3 backend Extbase plugins are registered with utility
functions of class :php:`\TYPO3\CMS\Extbase\Utility\ExtensionUtility` (not to

> be confused with :php:`\TYPO3\CMS\Core\Utility\ExtensionManagementUtility`).

Reviewer: THX

An Extbase plugin is configured for the frontend with
:php:`ExtensionUtility::configurePlugin()` in file
:file:`EXT:my_extension/ext_localconf.php`:

..  literalinclude:: _Plugins/_ext_localconf_extbase_plugin.php
    :caption: EXT:my_extension/ext_localconf.php

By using `ExtensionUtility::PLUGIN_TYPE_PLUGIN` as fifth parameter is is also

> possible to add the plugin as a list type. See :ref:`plugins-list_type`.

Reviewer: This is the default

Reviewer: Understand. You're using CONTENT_ELEMENT in the example and in futute CONTENT_ELEMENT will be the new default.

Author: at least so I hope. For the time being we want to suggest using CONTENT_ELEMENT without it being the default yet

Method :php:`ExtensionUtility::configurePlugin()` also takes care of registering
the plugin for frontend output in TypoScript using an object of type
:ref:`EXTBASEPLUGIN <t3tsref:cobj-extbaseplugin>`.

If it is desired that editors can insert the Extbase plugin like a content
element into the page it also needs to be registered with
:php:`ExtensionUtility::registerPlugin()` in the TCA Overrides, for example file
:file:`EXT:my_extension/Configuration/TCA/Overrides/tt_content.php`:

..  literalinclude:: _Plugins/_tt_content_extbase_plugin.php
    :caption: EXT:my_extension/Configuration/TCA/Overrides/tt_content.php

> For a detailed discussion of Extbase plugins including examples for controllers

Reviewer: ```suggestion
For a detailed explanation of Extbase plugins including examples for controllers
```

see chapter :ref:`extbase`.

..  _plugins-non-extbase:

Plugins without Extbase
~~~~~~~~~~~~~~~~~~~~~~~

It is possible to create a plugin without using Extbase by creating a plain PHP

> class as a controller.

Reviewer: Sorry. I have a problem with the "controller" wording. With "controller" I directly relate it to be part of the MVC pattern. It's a PHP class. And it's up to the coder to implement everything on his own or relate to the <out-dated> but still available AbstractPlugin class.

Author: An MVC pattern is not neccesarily bound to extbase. you can easily implement the controller pattern yourself.

Abstract plugin was deprecated with TYPO3 v12 and removed with TYPO3 v13 so we do not mention it here: https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.4/Deprecation-100639-DeprecateAbstractPlugin.html#deprecation-100639-deprecate-abstractplugin

In this case you have to define the TypoScript configuration yourself. A
:ref:`USER or USER_INT <t3tsref:cobj-user-int>` TypoScript object can be used
to delegate the rendering to your controller:

..  literalinclude:: _Plugins/plugin.typoscript
    :caption: EXT:my_extension/Configuration/TypoScript/setup.typoscript

To register such a plugin as content element you can use function

> :php:`ExtensionManagementUtility::addPlugin()` in the TCA overrides, for example

Reviewer: Somewhere above you have written about the pretty cool fifth argument to mark plugin as list-type. Why not using this fifth argument to mark plugin as content element? PLUGIN_TYPE_CONTENT_ELEMENT

Reviewer: Sorry. My fault. We are in Plugins without Extbase. Holy. Totally confused.

Author: Ack
